grails-app/services/org/pih/warehouse/data/ProductSynonymDataService.groovy

```diff
@@ -65,12 +68,16 @@ class ProductSynonymDataService {
             // Check if none error occures to be sure, that we have correct product, locale and synonymTypeCode
             // before we check the duplicates
             if (!command.errors.allErrors && synonymTypeCode == SynonymTypeCode.DISPLAY_NAME) {
-               Set<Synonym> duplicates = product?.synonyms?.findAll { Synonym synonym ->
-                   synonym.locale == new Locale(params['locale']) && synonym.synonymTypeCode == SynonymTypeCode.DISPLAY_NAME
-               }
-               if (duplicates) {
-                   command.errors.reject("Row ${index + 1}: You cannot add multiple display names in the same language. Edit the existing synonym instead.")
-               }
+                Set<Synonym> duplicates = product?.synonyms?.findAll { Synonym synonym ->
+                    synonym.locale == new Locale(params['locale']) && synonym.synonymTypeCode == SynonymTypeCode.DISPLAY_NAME
+                }
+                // If the product already has a synonym of type DISPLAY_NAME
+                // or we are trying to add it in any of the line above for this single import, throw a validation error
+                if (duplicates || productsWithDisplayName.find{ it == product?.productCode }) {
+                    command.errors.reject("Row ${index + 1}: You cannot add multiple display names in the same language. Edit the existing synonym instead.")
+                    return
+                }
+                productsWithDisplayName.add(product?.productCode)
```

Comment:
I think this also should be tied with `locale`, since you can have multiple display names in different locales.

Reply:
Oh yeah, that’s a bummer. Thanks for catching that